download/downloader.go

```diff
@@ -0,0 +1,536 @@
+	}
+
+	for fileIndex := range resumeState.CorruptedFiles {
+		for _, info := range fileInfoMap {
+			if info.Index == fileIndex {
+				logger.Warn("Removing corrupted file", "file", info.Record.FileName)
+				os.Remove(info.FullPath)
+				break
+			}
+		}
+	}
+
+	remainingBytes, remainingFiles := d.calculateRemaining(fileInfoMap, fileChunks, resumeState)
+	logger.Info("Resuming download",
+		"remaining", progress.FormatBytes(remainingBytes),
+		"files", remainingFiles)
+	logger.Info("Already downloaded",
+		"size", progress.FormatBytes(resumeState.BytesAlreadyDownloaded),
+		"files", resumeState.FilesAlreadyComplete)
+
+	return resumeState, nil
+}
+
+func (d *Downloader) calculateRemaining(fileInfoMap map[string]*FileInfo, fileChunks map[int][]manifest.ChunkRecord, resumeState *ResumeState) (int64, int) {
+	var remainingBytes int64
+	var remainingFiles int
+
+	for fileIndex := range fileChunks {
+		if resumeState.CompletedFiles[fileIndex] {
+			continue
+		}
+
+		for _, info := range fileInfoMap {
+			if info.Index != fileIndex {
+				continue
+			}
+			remainingFiles++
+			startChunk := resumeState.StartChunkIndex[fileIndex]
+			totalChunks := info.ChunkCount
+
+			for i := startChunk; i < totalChunks; i++ {
+				if i == totalChunks-1 {
+					lastChunkSize := int64(info.Record.SizeInBytes) - int64(totalChunks-1)*manifest.MaxChunkSize
+					if lastChunkSize > 0 {
+						remainingBytes += lastChunkSize
+					}
+				} else {
+					remainingBytes += manifest.MaxChunkSize
+				}
+			}
+			break
+		}
+	}
+
+	return remainingBytes, remainingFiles
+}
+
+func (d *Downloader) initializeProgress(fileInfoMap map[string]*FileInfo, resumeState *ResumeState) {
+	for _, info := range fileInfoMap {
+		chunksAlreadyWritten := 0
+		if resumeState != nil {
+			chunksAlreadyWritten = resumeState.StartChunkIndex[info.Index]
+		}
+		d.progress.AddFile(info.Index, info.Record.FileName, info.ChunkCount, int64(info.Record.SizeInBytes), chunksAlreadyWritten)
+	}
+
+	if resumeState != nil {
+		for fileIndex := range resumeState.CompletedFiles {
+			d.progress.FileComplete(fileIndex)
+		}
+		d.progress.AddDownloadedBytes(resumeState.BytesAlreadyDownloaded)
+	}
+}
+
+func (d *Downloader) downloadAndWrite(ctx context.Context, fileChunks map[int][]manifest.ChunkRecord, fileInfoMap map[string]*FileInfo, resumeState *ResumeState) error {
+	chunkJobs := make(chan ChunkJob, d.options.MaxDownloadWorkers*4)
+	downloadedChunks := make(chan ChunkResult, d.memory.maxMemory/manifest.MaxChunkSize)
+
+	fileIndexToInfo := make(map[int]*FileInfo)
+	fileIndexToChunks := make(map[int]int)
+	for _, info := range fileInfoMap {
+		fileIndexToInfo[info.Index] = info
+		fileIndexToChunks[info.Index] = len(fileChunks[info.Index])
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var downloadWg sync.WaitGroup
+
+	chunkDownloader := NewChunkDownloader(d.client, d.product, d.version.OS, d.progress)
+
+	for i := 0; i < d.options.MaxDownloadWorkers; i++ {
```

Comment:
Seems like we are correctly spawning the right amount of workers. Maybe we're making the previous channel too large